### ISSUE TITLE
Bugs in sphinx.util.osutil.relative_uri()

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#916]
sphinx.util.osutil.relative_uri turns paths into relative URIs.

For example 
    relative_uri('f0/f1/f2/source.html', 'f0/g1/g2/target.html')}}} returns {{{'../../g1/g2/target.html'}}}.
    {{{relative_uri('f0/f1/f2/source.html', 'f0/g1/g2/target.html')

Unfortunately, in the following four cases, relative_uri returns an incorrect result:

```
relative_uri('f0', 'f0/target.html') returns 'target.html', expected 'f0/target.html'
relative_uri('f0/source.html', 'f0/') returns '', expected './'
relative_uri('f0/f1/', 'f0/f1') returns '', expected '../f1'
relative_uri('f0/f1', 'f0/f1/') returns '', expected 'f1/'
```

This yields incorrect links in some case for HTML output.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/916
- Originally reported by: [macfreek](https://bitbucket.org/macfreek)
- Originally created at: 2012-04-28T19:45:59.397

**Comment by shimizukawa** [comment]
_From [macfreek](https://bitbucket.org/macfreek) on 2012-04-28 19:11:20+00:00_

See attached files for the old and new code and some unit tests.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2012-10-28 17:13:40+00:00_

relative_uri patch. Fixes #916.

→ <<cset 2f1f4c6be2dfc5caec235d4309d948b2a87bde8a>>
